runtime/signals_nat.c

```diff
@@ -185,37 +179,29 @@ DECLARE_SIGNAL_HANDLER(trap_handler)
 
 #ifdef HAS_STACK_OVERFLOW_DETECTION
 
-static char * system_stack_top;
 static char sig_alt_stack[SIGSTKSZ];
 
-#if defined(SYS_linux)
-/* PR#4746: recent Linux kernels with support for stack randomization
-   silently add 2 Mb of stack space on top of RLIMIT_STACK.
-   2 Mb = 0x200000, to which we add 8 kB (=0x2000) for overshoot. */
-#define EXTRA_STACK 0x202000
-#else
-#define EXTRA_STACK 0x2000
-#endif
+/* Code compiled with ocamlopt never accesses more than
+   EXTRA_STACK bytes below the stack pointer. */
+#define EXTRA_STACK 256
 
 #ifdef RETURN_AFTER_STACK_OVERFLOW
 extern void caml_stack_overflow(void);
 #endif
 
 DECLARE_SIGNAL_HANDLER(segv_handler)
 {
-  struct rlimit limit;
   struct sigaction act;
   char * fault_addr;
 
   /* Sanity checks:
      - faulting address is word-aligned
-     - faulting address is within the stack
+     - faulting address is on the stack, or within EXTRA_STACK of it
      - we are in OCaml code */
   fault_addr = CONTEXT_FAULTING_ADDRESS;
   if (((uintnat) fault_addr & (sizeof(intnat) - 1)) == 0
-      && getrlimit(RLIMIT_STACK, &limit) == 0
-      && fault_addr < system_stack_top
-      && fault_addr >= system_stack_top - limit.rlim_cur - EXTRA_STACK
+      && fault_addr < caml_top_of_stack
```

Comment:
I'm afraid `caml_top_of_stack` is valid only for the initial thread and makes no sense for other threads.

I though the heuristic you proposed was "faulting address is close enough to the stack pointer", without any reference to a "top of stack" address.

Reply:
> I'm afraid caml_top_of_stack is valid only for the initial thread and makes no sense for other threads.

I also thought so, but yesterday discovered that `st_stubs.c` contains code in `caml_thread_{save,restore}_runtime_state` to keep `caml_top_of_stack` updated and relevant to the current thread.

> I though the heuristic you proposed was "faulting address is close enough to the stack pointer", without any reference to a "top of stack" address.

That was indeed what I initially proposed. I'm happy to use that heurstic instead of caml_top_of_stack, but since it seems to be maintained for each thread I thought we may as well use it. Are there further accuracy problems with caml_top_of_stack?

Reply:
For reference: `caml_top_of_stack` was fixed by #996

Reply:
You're right, I misremembered / mis-grepped the handling of `caml_top_of_stack`.  We need to maintain this variable in order to compute the stack usage part of GC stats, so why not use it, indeed. 

I think its value is reasonably accurate -- up to the size of the frame size of a small C runtime function, that's not much uncertainty.  The way `caml_top_of_stack` is initialized is very much undefined C behavior, but that hasn't been a problem so far, and could be changed later.